server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -993,14 +993,15 @@ public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
             String name,
             boolean updateable,
             Function<FieldMapper, T> initializer,
-            T defaultValue,
+            Supplier<T> defaultValue,
```

Comment:
We need consistency between all our mapper default values. I do not like having a specialized constructor here that takes a supplier but none of the others do.

I need to think a bit more. I would prefer not having supplied default value like this.

Reply:
I understand. The reasons for adding this were:
- We need to change the default value for similarity based on the value for `index`.
- Other mapper builders provide this as well (indexParam / boolParam, analyzerParam).

Are there better ways to achieve this? I can give it a try 👍

Reply:
> Other mapper builders provide this as well (indexParam / boolParam, analyzerParam).

As long as we are unified with the builders, it could be ok. I will read them. My concern is that somebody passes "null" and instead of it being "() -> null" is `null` and we call `get()` on `null`

Reply:
Yep, that happened 😢 . I added an `assert` to try and catch that as early as possible, but this could definitely bite us.

Reply:
@carlosdelest I would suggest a NEW method/ctor that works with enums that has a supplier. This way if somebody is passing `null`, its a compilation error as Java isn't sure which one you are specifying. That is way better than a runtime error.

Reply:
Good call - adding that. Now we need to specify what method are we using to avoid surprises. Thanks for the suggestion! 🙇